### PR TITLE
Cleanarchitecture

### PR DESCRIPTION
**Pull request by andresledererdobra**
Clean architecture with 3 layers : app, domain, data

### CODE REVIEW

**Comment by andresledererdobra** [comment]
Screenshot (22)

clean architecture structure

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -10,14 +10,15 @@
...
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
 
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <activity android:name=".Viewmodel"/>
     </application>
 
 </manifest>
```
Comment (SartiMau) [inline app/src/main/AndroidManifest.xml]: to avoid this error just add an empty line at the end of the file pls

File: build.gradle [hunk 2]
```diff
@@ -15,4 +15,23 @@ buildscript {
...
+        testLogging {
+            exceptionFormat "full"
+            showCauses true
+            showExceptions true
+            showStackTraces true
+            showStandardStreams true
+            events = ["passed", "skipped", "failed", "standardOut", "standardError"]
+        }
+    }
 }
```
Comment (SartiMau) [inline build.gradle]: same here

File: data/build.gradle [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+dependencies {
+    implementation project(path: ':domain')
+    implementation 'androidx.core:core-ktx:1.7.0'
+    implementation 'androidx.appcompat:appcompat:1.4.1'
+    implementation 'com.google.android.material:material:1.5.0'
+    testImplementation 'junit:junit:4.13.2'
+    androidTestImplementation 'androidx.test.ext:junit:1.1.3'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
+}
```
Comment (SartiMau) [inline data/build.gradle]: same here

File: data/src/androidTest/java/com/example/data/ExampleInstrumentedTest.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.example.data
+
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.ext.junit.runners.AndroidJUnit4
+
+import org.junit.Test
+import org.junit.runner.RunWith
+
+import org.junit.Assert.*
```
Comment (SartiMau) [inline data/src/androidTest/java/com/example/data/ExampleInstrumentedTest.kt]: do not use wildcard imports pls

File: data/src/test/java/com/example/data/ExampleUnitTest.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.example.data
+
+import org.junit.Test
+
+import org.junit.Assert.*
```
Comment (SartiMau) [inline data/src/test/java/com/example/data/ExampleUnitTest.kt]: do not use wildcard imports

File: data/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.example.data">
+
+</manifest>
```
Comment (SartiMau) [inline data/src/main/AndroidManifest.xml]: add an empty line at the end of the file pls

File: data/src/test/java/com/example/data/ExampleUnitTest.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
+    @Test
+    fun addition_isCorrect() {
+        assertEquals(4, 2 + 2)
+    }
+}
```
Comment (SartiMau) [inline data/src/test/java/com/example/data/ExampleUnitTest.kt]: add an empty line at the end of the file pls

File: di/build.gradle [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+dependencies {
+    implementation project(path:':domain')
+    implementation project(path:':data')
+    implementation 'androidx.core:core-ktx:1.7.0'
+    implementation 'androidx.appcompat:appcompat:1.4.1'
+    implementation 'com.google.android.material:material:1.5.0'
+    testImplementation 'junit:junit:4.13.2'
+    androidTestImplementation 'androidx.test.ext:junit:1.1.3'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
+}
```
Comment (SartiMau) [inline di/build.gradle]: add an empty line at the end of the file pls

File: di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ */
+@RunWith(AndroidJUnit4::class)
+class ExampleInstrumentedTest {
+    @Test
+    fun useAppContext() {
+        // Context of the app under test.
+        val appContext = InstrumentationRegistry.getInstrumentation().targetContext
+        assertEquals("com.example.di.test", appContext.packageName)
+    }
+}
```
Comment (SartiMau) [inline di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt]: add an empty line at the end of the file pls

File: di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package com.example.di
+
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.ext.junit.runners.AndroidJUnit4
+
+import org.junit.Test
+import org.junit.runner.RunWith
+
+import org.junit.Assert.*
```
Comment (SartiMau) [inline di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt]: do not use wildcard imports pls

File: di/src/test/java/com/example/di/ExampleUnitTest.kt [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package com.example.di
+
+import org.junit.Test
+
+import org.junit.Assert.*
```
Comment (SartiMau) [inline di/src/test/java/com/example/di/ExampleUnitTest.kt]: do not use wildcard imports pls

File: di/src/test/java/com/example/di/ExampleUnitTest.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
+    @Test
+    fun addition_isCorrect() {
+        assertEquals(4, 2 + 2)
+    }
+}
```
Comment (SartiMau) [inline di/src/test/java/com/example/di/ExampleUnitTest.kt]: add an empty line at the end of the file pls

File: domain/build.gradle [hunk 13]
```diff
@@ -0,0 +1,9 @@
+plugins {
+    id 'java-library'
+    id 'org.jetbrains.kotlin.jvm'
+}
+
+java {
+    sourceCompatibility = JavaVersion.VERSION_1_7
+    targetCompatibility = JavaVersion.VERSION_1_7
+}
```
Comment (SartiMau) [inline domain/build.gradle]: add an empty line at the end of the file pls

File: domain/src/main/java/com/example/domain/Domain.kt [hunk 14]
```diff
@@ -0,0 +1,4 @@
+package com.example.domain
+
+class Domain {
+}
```
Comment (SartiMau) [inline domain/src/main/java/com/example/domain/Domain.kt]: add an empty line at the end of the file pls

**Review by SartiMau — CHANGES_REQUESTED** [review]
1. Remove all the .idea files pls
2. Review because it is clear that something happened with the ktlint in more than one module because it is not working fine

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (andresledererdobra) [reply app/src/main/AndroidManifest.xml]: checked!

File: build.gradle (hunk 2 above)
Comment (andresledererdobra) [reply build.gradle]: checked

File: data/build.gradle (hunk 3 above)
Comment (andresledererdobra) [reply data/build.gradle]: done!

File: data/src/androidTest/java/com/example/data/ExampleInstrumentedTest.kt (hunk 4 above)
Comment (andresledererdobra) [reply data/src/androidTest/java/com/example/data/ExampleInstrumentedTest.kt]: done!

File: data/src/test/java/com/example/data/ExampleUnitTest.kt (hunk 5 above)
Comment (andresledererdobra) [reply data/src/test/java/com/example/data/ExampleUnitTest.kt]: done!

File: data/src/main/AndroidManifest.xml (hunk 6 above)
Comment (andresledererdobra) [reply data/src/main/AndroidManifest.xml]: done!

File: data/src/test/java/com/example/data/ExampleUnitTest.kt (hunk 7 above)
Comment (andresledererdobra) [reply data/src/test/java/com/example/data/ExampleUnitTest.kt]: done!

File: di/build.gradle (hunk 8 above)
Comment (andresledererdobra) [reply di/build.gradle]: done!

File: di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt (hunk 9 above)
Comment (andresledererdobra) [reply di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt]: done!

File: di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt (hunk 10 above)
Comment (andresledererdobra) [reply di/src/androidTest/java/com/example/di/ExampleInstrumentedTest.kt]: done!

File: di/src/test/java/com/example/di/ExampleUnitTest.kt (hunk 11 above)
Comment (andresledererdobra) [reply di/src/test/java/com/example/di/ExampleUnitTest.kt]: done!

File: di/src/test/java/com/example/di/ExampleUnitTest.kt (hunk 12 above)
Comment (andresledererdobra) [reply di/src/test/java/com/example/di/ExampleUnitTest.kt]: done!

File: domain/build.gradle (hunk 13 above)
Comment (andresledererdobra) [reply domain/build.gradle]: done!

File: domain/src/main/java/com/example/domain/Domain.kt (hunk 14 above)
Comment (andresledererdobra) [reply domain/src/main/java/com/example/domain/Domain.kt]: done!

File: data/src/androidTest/java/com/example/data/ExampleInstrumentedTest.kt [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+ */
+@RunWith(AndroidJUnit4::class)
+class ExampleInstrumentedTest {
+    @Test
+    fun useAppContext() {
+        // Context of the app under test.
+        val appContext = InstrumentationRegistry.getInstrumentation().targetContext
+        assertEquals("com.example.data.test", appContext.packageName)
+    }
+}
```
Comment (SartiMau) [inline data/src/androidTest/java/com/example/data/ExampleInstrumentedTest.kt:24]: add an empty line here to avoid this error

**Review by SartiMau — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]

**Review by mariano98montero — APPROVED** [review]

**Pull request merged**